drape/text_itemizer.cpp

```diff
@@ -0,0 +1,270 @@
+{
+  // Fill scripts with the script extensions.
+  UErrorCode icu_error = U_ZERO_ERROR;
+  size_t const count = uscript_getScriptExtensions(codepoint, scripts, kMaxScripts, &icu_error);
+  if (U_FAILURE(icu_error))
+    return 0;
+
+  return count;
+}
+
+// Intersects the script extensions set of codepoint with result and writes to result, reading and updating resultSize.
+// The output result will be a subset of the input result (thus resultSize can only be smaller).
+void ScriptSetIntersect(UChar32 codepoint, UScriptCode * result, size_t * resultSize)
+{
+  // Each codepoint has a Script property and a Script Extensions (Scx) property.
+  //
+  // The implicit Script property values 'Common' and 'Inherited' indicate that a codepoint is widely used in many
+  // scripts, rather than being associated to a specific script.
+  //
+  // However, some codepoints that are assigned a value of 'Common' or 'Inherited' are not commonly used with all
+  // scripts, but rather only with a limited set of scripts. The Script Extension property is used to specify the set
+  // of script which borrow the codepoint.
+  //
+  // Calls to GetScriptExtensions(...) return the set of scripts where the codepoints can be used.
+  // (see table 7 from http://www.unicode.org/reports/tr24/tr24-29.html)
+  //
+  //     Script       Script Extensions ->  Results
+  //  1) Common       {Common}          ->  {Common}
+  //     Inherited    {Inherited}       ->  {Inherited}
+  //  2) Latin        {Latn}            ->  {Latn}
+  //     Inherited    {Latn}            ->  {Latn}
+  //  3) Common       {Hira Kana}       ->  {Hira Kana}
+  //     Inherited    {Hira Kana}       ->  {Hira Kana}
+  //  4) Devanagari   {Deva Dogr Kthi Mahj}  ->  {Deva Dogr Kthi Mahj}
+  //     Myanmar      {Cakm Mymr Tale}  ->  {Cakm Mymr Tale}
+  //
+  // For most of the codepoints, the script extensions set contains only one element. For CJK codepoints, it's common
+  // to see 3-4 scripts. For really rare cases, the set can go above 20 scripts.
+  UScriptCode scripts[kMaxScripts] = { USCRIPT_INVALID_CODE };
+  size_t const count = GetScriptExtensions(codepoint, scripts);
+
+  // Implicit script 'inherited' is inheriting scripts from preceding codepoint.
+  if (count == 1 && scripts[0] == USCRIPT_INHERITED)
+    return;
+
+  auto const contains = [&scripts, count](UScriptCode code)
+  {
+    for (size_t i = 0; i < count; ++i)
+      if (scripts[i] == code)
+        return true;
+
+    return false;
+  };
+
+  // Perform the intersection of both script set.
+  ASSERT(!contains(USCRIPT_INHERITED), ());
+  size_t outSize = 0;
+  for (size_t i = 0; i < *resultSize; ++i)
+  {
+    auto const current = result[i];
+    if (contains(current))
+      result[outSize++] = current;
+  }
+
+  *resultSize = outSize;
+}
+
+// Iterates through the characters in UTF-16 string. Example:
+// for (UTF16CharIterator iter(str); !iter.end(); iter.Next())
+//   std::cout << iter.get();
+class UTF16CharIterator
```

Comment:
I suppose it should already present in 3party utf8 like utf8::unchecked::iterator ?

Reply:
This one is a bit different, check how it is used in the code. It requires a bit of tweaks. The right one is `next16` method. Any help is appreciated )

Reply:
Rewrote the code and removed this iterator.